src/hamiltonian/non_local_operator.cpp

```diff
@@ -414,67 +418,79 @@ void
+        case device_t::CPU: {
+            la = linalg_t::blas;
+            mt = memory_t::host;
+            break;
+        }
+        case device_t::GPU: {
+            la = linalg_t::gpublas;
+            mt = memory_t::device;
+            break;
+        }
+        default:
+            break;
+    }
     /* First calculate the local part of the projections
        dm(i, n) = <phi_i| S |psi_{nk}> */
     sddk::inner(ctx__.spla_context(), spins__, hub_wf__, 0, hub_wf__.num_wf(), phi__, N__, n__, dm, 0, 0);
 
     dmatrix<std::complex<T>> Up(hub_wf__.num_wf(), n__);
-    Up.zero();
-
-    //auto nhwf = um__.nhwf();
-
-    #pragma omp parallel for schedule(static)
-    for (int ia = 0; ia < ctx__.unit_cell().num_atoms(); ia++) {
-        auto& atom = ctx__.unit_cell().atom(ia);
-        if (atom.type().hubbard_correction()) {
-            const int lmax_at = 2 * atom.type().indexr_hub().am(0).l() + 1;
-            // we apply the hubbard correction. For now I have no papers
-            // giving me the formula for the SO case so I rely on QE for it
-            // but I do not like it at all
-            if (ctx__.num_mag_dims() == 3) {
+    if (ctx__.processing_unit() == device_t::GPU) {
+        Up.allocate(memory_t::device);
+    }
+
+    if (ctx__.num_mag_dims() == 3) {
+        Up.zero();
+        #pragma omp parallel for schedule(static)
+        for (int at_lvl = 0; at_lvl < (int)um__.atomic_orbitals().size(); at_lvl++) {
+            const int ia     = um__.atomic_orbitals(at_lvl).first;
+            auto const& atom = ctx__.unit_cell().atom(ia);
+            if (atom.type().lo_descriptor_hub(um__.atomic_orbitals(at_lvl).second).use_for_calculation()) {
+                const int lmax_at = 2 * atom.type().lo_descriptor_hub(um__.atomic_orbitals(at_lvl).second).l() + 1;
+                // we apply the hubbard correction. For now I have no papers
+                // giving me the formula for the SO case so I rely on QE for it
+                // but I do not like it at all
                 for (int s1 = 0; s1 < ctx__.num_spins(); s1++) {
                     for (int s2 = 0; s2 < ctx__.num_spins(); s2++) {
-                        const int ind = (s1 == s2) * s1 + (1 + 2 * s2 + s1) * (s1 != s2);
-
-                        // TODO: replace this with matrix matrix multiplication
-
+                      // TODO: replace this with matrix matrix multiplication
                         for (int nbd = 0; nbd < n__; nbd++) {
                             for (int m1 = 0; m1 < lmax_at; m1++) {
                                 for (int m2 = 0; m2 < lmax_at; m2++) {
-                                    Up(um__.nhwf() * s1 + um__.offset(ia) + m1, nbd) +=
-                                        um__(um__.offset(ia) + m2, um__.offset(ia) + m1, ind) *
-                                        dm(um__.nhwf() * s2 + um__.offset(ia) + m2, nbd);
+                                    const int ind = (s1 == s2) * s1 + (1 + 2 * s2 + s1) * (s1 != s2);
```

Comment:
It is probably worth to have a small macro or inlned function the sirius namespace for computing these indices. It is not blocking though and can wait.